five_tips_about_testable_golang_code.md

```diff
@@ -0,0 +1,171 @@
+# How to write untestable code in golang
+According to my experience in movies, if you want to be a super bad guy, you will need to work harder than the good guys. Think about it! You need to hire evil scientists, build huge death bomb to threaten the world, a good plan and a lot of money. But be a good guy basically just need some brave hearts, you do not even have known about the structure of bomb because when choosing red/blue wire to cut, good guys never mess up.
+
+Yeah, that's not fair.
+
+Good news (?) everyone, in software developer side, mess up code is far easier than build it in a good way. If you keep following points in mind, make golang code untestable would be so easy.
+
+Before reading this, I assume you already read [Aaron's Blog: Testify - A Toolkit for Golang Unit Tests](https://vendasta-blog.appspot.com/blog/BL-F5QKDT6H/).
+
+## 1. Always expected golang mock as powerful as Python mock
+
+This is the easiest part. If you start with something with Python, all you need is ~~love~~ do nothing extra, just write code with all Python style, you can make it untestable.
+
+First, let take a look at this chart:
+
+|        | python mock | golang testify mock |
+|--------|-------------|-------------|
+|mock a class|  Yes|         Yes    |
+|mock a method in a mock class| Yes |Yes|
+|assert mock been called with given parameter | Yes | Yes|
+|return any result in mock method | Yes |Yes|
+|mock an attribute in a mock class | Yes | No |
+|mock a function | Yes | No|
+| patch | Yes | No |
+
+![that is how good Python is.](https://imgs.xkcd.com/comics/python.png)
+
+Mock(testify) in golang is good. Just not that *magic* as Python's. When I first learn how to use it, the thought occurred to my mind most frequently is "What, I need to do this manually?" and "If I need to do this manually, what are YOU doing, mock?"
```

Comment:
"Mock (Testify) in Golang is good, just not quite as *magical* as mocking in Python. When I first learned how to use it, the thought that occurred to me most frequently was ..."